lib/msf/core/handler/bind_aws_instance_connect.rb

```diff
@@ -0,0 +1,347 @@
+    end
+
+    # Start a new handler
+    start_handler
+  end
+
+  #
+  # Starts monitoring for an outbound connection to become established.
+  #
+  def start_handler
+
+    # Maximum number of seconds to run the handler
+    ctimeout = 150
+
+    # Maximum number of seconds to await initial API response
+    rtimeout = 5
+
+    if (exploit_config and exploit_config['active_timeout'])
+      ctimeout = exploit_config['active_timeout'].to_i
+    end
+    self.listener_pairs[datastore['EC2_ID']] = true
+
+    # Start a new handling thread
+    self.listener_threads = framework.threads.spawn("BindAwsInstanceConnectHandler-#{datastore['EC2_ID']}", false) {
+      instance_connect_client = nil
+
+      print_status("Started #{human_name} handler against #{datastore['EC2_ID']}:#{datastore['REGION']}")
+
+      if (datastore['EC2_ID'] == nil or datastore['EC2_ID'].strip.empty?)
+        raise ArgumentError,
+          "EC2_ID is not defined; InstanceConnect handler cannot function.",
+          caller
+      end
+
+      stime = Time.now.to_i
+
+      while (stime + ctimeout > Time.now.to_i)
+        begin
+          # Call API to start InstanceConnect session
+          if start_instance_connect_session
+            instance_connect_client = connect_ssh
+          else
+            raise Rex::ConnectionError.new('Cannot establish serial connection to ' + datastore['EC2_ID'])
+          end
+        rescue Rex::ConnectionError => e
+          vprint_error(e.message)
+        rescue
+          wlog("Exception caught in InstanceConnect handler: #{$!.class} #{$!}")
+          break
+        end
+        break if instance_connect_client
+
+        # Wait a second before trying again
+        Rex::ThreadSafe.sleep(0.5)
+      end
+
+      # Valid client connection?
+      if (instance_connect_client)
+        # Increment the has connection counter
+        self.pending_connections += 1
+
+        # Timeout and datastore options need to be passed through to the client
+        opts = {
+          :datastore    => datastore,
+          :expiration   => datastore['SessionExpirationTimeout'].to_i,
+          :comm_timeout => datastore['SessionCommunicationTimeout'].to_i,
+          :retry_total  => datastore['SessionRetryTotal'].to_i,
+          :retry_wait   => datastore['SessionRetryWait'].to_i,
+        }
+
+        self.conn_threads << framework.threads.spawn("BindAwsInstanceConnectHandlerSession", false, instance_connect_client, opts) { |ssh, opts_copy|
+          begin
+            self.listener_pairs[datastore['EC2_ID']] = ssh
+            handle_connection(ssh, opts_copy)
+          rescue => e
+            elog('Exception raised from BindAwsInstanceConnect.handle_connection', error: e)
+          end
+        }
+      else
+        wlog("No connection received before the handler completed")
+      end
+    }
+  end
+
+  # A URI describing what the payload is configured to use for transport
+  def payload_uri
+    "ssm://#{datastore['EC2_ID']}:0"
```

Comment:
Should this still be ssm?

Reply:
No, no it shouldnt.
How's `serial+ssh://` sound?